apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java

```diff
@@ -0,0 +1,140 @@
+package com.ctrip.framework.apollo.core.utils;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.slf4j.Logger;
+import org.slf4j.event.Level;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Delayed log printing utility class, used only for logging when Apollo is initialized
+ *
+ * @author kl (http://kailing.pub)
+ * @since 2021/5/11
+ */
+final class DeferredLogCache {
+
+  public static final int MAX_LOG_SIZE = 1000;
+  private static final AtomicInteger LOG_INDEX = new AtomicInteger(0);
+  private static final Cache<Integer, Line> LOG_CACHE = CacheBuilder.newBuilder()
+      .maximumSize(MAX_LOG_SIZE)
+      .expireAfterWrite(1, TimeUnit.MINUTES)
+      .build();
+
+  private DeferredLogCache() {
+  }
+
+
+  public static void debug(Logger logger, String message, Object... objects) {
+    add(logger, Level.DEBUG, message, objects, null);
+  }
+
+  public static void info(Logger logger, String message, Object... objects) {
+    add(logger, Level.INFO, message, objects, null);
+  }
+
+  public static void warn(Logger logger, String message, Object... objects) {
+    add(logger, Level.WARN, message, objects, null);
+  }
+
+  public static void error(Logger logger, String message, Throwable throwable) {
+    add(logger, Level.ERROR, message, null, throwable);
+  }
+
+  private static void add(Logger logger, Level level, String message, Object[] objects,
+      Throwable throwable) {
+    Line logLine = new Line(level, message, objects, throwable, logger);
+    LOG_CACHE.put(LOG_INDEX.incrementAndGet(), logLine);
+  }
+
+  public static void replayTo() {
+    if (DeferredLogger.isEnabled()) {
+      DeferredLogger.disable();
+      for (int i = 1; i <= LOG_INDEX.get(); i++) {
+        Line logLine = LOG_CACHE.getIfPresent(i);
+        assert logLine != null;
+        Logger logger = logLine.getLogger();
+        Level level = logLine.getLevel();
+        String message = logLine.getMessage();
+        Object[] objects = logLine.getObjects();
+        Throwable throwable = logLine.getThrowable();
+        logTo(logger, level, message, objects, throwable);
+      }
+      clear();
+    }
+
+  }
```

Comment:
I think DeferredLogCache could remain simple by not caring whether DeferredLogger is enabled or not.
Actually it could even be renamed to LogCache which means it's simple a cache.
```suggestion
  public static void replayTo() {
    for (int i = 1; i <= LOG_INDEX.get(); i++) {
      Line logLine = LOG_CACHE.getIfPresent(i);
      assert logLine != null;
      Logger logger = logLine.getLogger();
      Level level = logLine.getLevel();
      String message = logLine.getMessage();
      Object[] objects = logLine.getObjects();
      Throwable throwable = logLine.getThrowable();
      logTo(logger, level, message, objects, throwable);
    }
    clear();
  }
```

Reply:
yes,it is